In [8]:
#import libraries
import pandas as pd
import numpy as np
import os

In [18]:
#create path
path= r"C:\Users\beaac\OneDrive\Documents\Instacart Basket 25"




In [20]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path,'02 data' , 'prepared data' , 'ords_prods_merge.pkl'))


In [26]:
# Subset of df
df= df_ords_prods_merge[:1000000]

In [34]:
# set the conditions 
# defining the function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [36]:
# apply the function
df[ 'price_range'] = df.apply(price_label, axis=1)

C:\Users\beaac\AppData\Local\Temp\ipykernel_4412\52463990.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[ 'price_range'] = df.apply(price_label, axis=1)


In [38]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673183
Low-range product    314392
High range            12425
Name: count, dtype: int64

In [40]:
df.shape

(1000000, 17)

In [42]:
df['prices'].max()

99999.0

USING IF-STATEMENTS WITH LOC() FUNCTION 

In [80]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'prices_range_loc'] = 'High-range product'

In [88]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] >5), 'prices_range_loc'] = 'Mid-range product'

In [84]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 15, 'prices_range_loc'] = 'Low-range product'

In [90]:
df_ords_prods_merge['prices_range_loc'].value_counts(dropna= False)


prices_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [54]:
df_ords_prods_merge.shape

(32404859, 18)

IF STATEMENTS WITH FOR LOOPS

In [56]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [93]:
#Uing for-loop to create bussiest day column
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [95]:
#combiing ords_prods_merge df with busiest day
df_ords_prods_merge['busiest_day'] =result

In [64]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# TASK 4.7


CREATE HE COLUMN "BUSIEST DAYS' WITH TWO BUSIEST DAYS AND TWO SLOWEST DAYS

In [114]:
#CREATE COLUMN BUSIEST DAYS 
#IDENTIFY TWO BUSIEST DAYS (0 and 1)
# IDENTIFY TWO SLOWEST DAYS (4 and 3)

result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value ==1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3 :
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")


In [115]:
result_2

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

CHECK VALUES OF NEW BUSIEST DAYS FOR ACCURACY 

In [118]:

#combiing ords_prods_merge df with busiest dayS
df_ords_prods_merge['busiest_days'] = result_2

In [121]:
#EXECUTE RESULTS
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [123]:
#check ords_prods_merge with new busiest days column
df_ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  8                     0.0       14084                  2   
2                  8                     0.0       12427                  3   
3                  8                     0.0       26088                  4   
4                  8                     0.0       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match   prices_range_loc    price_range_loc  \
0              7     9.0  both  Mid-range product  Low-range product   
1             16    12.5  both  Mid-range product  Low-range product   
2             19     4.4  both  Low-range product  Low-range product   
3             19     4.7  both  Low-range product  Low-range product   
4             17     1.0  both  Low-range product  Low-range product   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy  Regularly busy  
2  Regularly busy  Regularly busy  
3  Regularly busy  Regularly busy  
4  Regularly busy  Regularly busy

The total value counts for the ord_prods_merge dataframe is equal to the total sum of all grouping (regularly busy+ busiest days+ slowest days). each label match the sum totalts for the qualifying days of the week like for example busiest days =11864412 , witch match orders day of week 0+1 .

BUSIEST HOURS OF THE DAY - "MOST ORDERS", "AVERAGE ORDERS" AND "FEWEST ORDERS"

In [128]:
# check value counts in order hour of day column
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

I split the list in equal thirds:
Most orders: 10,11,14,15,13,12,16,9
Fewest orders: 23,6,0,1,5,2,4,3
Average orders: All the remaining values


In [131]:
#Uing for-loop to create bussiest period of day
result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10,11,14,15,13,12,16,9]:
    result_3.append("Most orders")
  elif value in [23,6,0,1,5,2,4,3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [133]:
result_3

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera


CHECK VALUES OF NEW BUSIEST PERIOD OF DAY FOR ACCURACY

In [136]:
#combiing ords_prods_merge df with Busiest periods of day
df_ords_prods_merge['busiest_period_of_day'] = result_3

In [140]:
#Print frequency of busiest period of day
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [144]:
#export dtaframe as pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))